library/pkcs7.c

```diff
@@ -0,0 +1,525 @@
+ **/
+static int pkcs7_get_signers_info_set( unsigned char **p, unsigned char *end,
+                                       mbedtls_pkcs7_signer_info *signers_set )
+{
+    unsigned char *end_set;
+    int ret;
+    size_t len = 0;
+
+    ret = mbedtls_asn1_get_tag( p, end, &len, MBEDTLS_ASN1_CONSTRUCTED
+            | MBEDTLS_ASN1_SET );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO + ret );
+
+    end_set = *p + len;
+
+    ret = mbedtls_asn1_get_tag( p, end_set, &len, MBEDTLS_ASN1_CONSTRUCTED
+            | MBEDTLS_ASN1_SEQUENCE );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO + ret );
+
+    ret = mbedtls_asn1_get_int( p, end_set, &signers_set->version );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO + ret );
+
+    ret = mbedtls_asn1_get_tag( p, end_set, &len, MBEDTLS_ASN1_CONSTRUCTED
+            | MBEDTLS_ASN1_SEQUENCE );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO + ret );
+
+    signers_set->issuer_raw.p = *p;
+
+    ret = mbedtls_asn1_get_tag( p, end_set, &len, MBEDTLS_ASN1_CONSTRUCTED
+            | MBEDTLS_ASN1_SEQUENCE );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO + ret );
+
+    ret  = mbedtls_x509_get_name( p, *p + len, &signers_set->issuer );
+    if( ret != 0 )
+        return( ret );
+
+    signers_set->issuer_raw.len =  *p - signers_set->issuer_raw.p;
+
+    ret = mbedtls_x509_get_serial( p, end_set, &signers_set->serial );
+    if( ret != 0 )
+        return( ret );
+
+    ret = pkcs7_get_digest_algorithm( p, end_set,
+            &signers_set->alg_identifier );
+    if( ret != 0 )
+        return( ret );
+
+    ret = pkcs7_get_digest_algorithm( p, end_set,
+            &signers_set->sig_alg_identifier );
+    if( ret != 0 )
+        return( ret );
+
+    ret = pkcs7_get_signature( p, end, &signers_set->sig );
+    if( ret != 0 )
+        return( ret );
+
+    signers_set->next = NULL;
+
+    return( 0 );
+}
+
+/**
+ * SignedData ::= SEQUENCE {
+ *      version Version,
+ *      digestAlgorithms DigestAlgorithmIdentifiers,
+ *      contentInfo ContentInfo,
+ *      certificates
+ *              [0] IMPLICIT ExtendedCertificatesAndCertificates
+ *                  OPTIONAL,
+ *      crls
+ *              [0] IMPLICIT CertificateRevocationLists OPTIONAL,
+ *      signerInfos SignerInfos }
+ */
+static int pkcs7_get_signed_data( unsigned char *buf, size_t buflen,
+        mbedtls_pkcs7_signed_data *signed_data )
+{
+    unsigned char *p = buf;
+    unsigned char *end = buf + buflen;
+    size_t len = 0;
+    int ret;
+    mbedtls_md_type_t md_alg;
+
+    ret = mbedtls_asn1_get_tag( &p, end, &len, MBEDTLS_ASN1_CONSTRUCTED
+            | MBEDTLS_ASN1_SEQUENCE );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_FORMAT + ret );
+
+    /* Get version of signed data */
+    ret = pkcs7_get_version( &p, end, &signed_data->version );
```

Comment:
Blocker: `end` needs to be adjusted to the end of the `SEQUENCE` here. Otherwise, as above, the content of the `SignerInfo` can potentially overrun the declared bounds of the latter.